ordered_product/models.py

```diff
@@ -0,0 +1,105 @@
+    password = models.CharField(max_length=16,
+                                validators=[MinLengthValidator(6)],
+                                null=False,
+                                help_text="Password")
+
+    class Meta:
+        abstract = True
+
+
+class Supplier(User):
+    business_name = models.CharField(max_length=32,
+                                     validators=[MinLengthValidator(6)],
+                                     null=False,
+                                     help_text="Business Name")
+
+
+class DeliveryLocation(models.Model):
+    user_name = models.ForeignKey(Supplier, on_delete=models.CASCADE)
+    location = models.CharField(max_length=32, validators=[MinLengthValidator(1)])
+    date = models.DateField()
+
+
+class Client(User):
+    area = models.CharField(max_length=32,
+                            validators=[MinLengthValidator(6)],
+                            help_text="area")
+
+
+class SupplierProduct(models.Model):
+    suppler_product_id = models.IntegerField(primary_key=True)
+    qr_code = models.ForeignKey(Product, on_delete=models.CASCADE)
+    user_name = models.ForeignKey(Supplier, on_delete=models.CASCADE)
+    price = models.IntegerField(validators=[MinValueValidator(1)])
+    quantity = models.IntegerField(validators=[MinValueValidator(1)])
+
+    class Meta:
+        unique_together = ('suppler_product_id', 'qr_code')
+
+
+class OrderedProduct(models.Model):
+    delivery_location_id = models.ForeignKey(DeliveryLocation, on_delete=models.CASCADE)
+    user_name = models.ForeignKey(Client, on_delete=models.CASCADE)
+    supplier_product_id = models.ForeignKey(SupplierProduct, on_delete=models.CASCADE)
+    quantity = models.IntegerField(default=1, validators=[MinValueValidator(1)])
+
+    @staticmethod
+    def filter_id(opID):
+        return OrderedProduct.objects.filter(id=opID) if opID else None
+
+    @staticmethod
+    def filter_user_name(userName):
+        return OrderedProduct.objects.filter(user_name=userName) if userName else []
+
+    @staticmethod
+    def filter_delivery_location_id(dlID):
+        return OrderedProduct.objects.filter(delivery_location_id=dlID) if dlID else []
+
+    @staticmethod
+    def filter_supplier_product_id(spID):
+        return OrderedProduct.objects.filter(supplier_product_id=spID) if spID else []
+
+    def save_ordered_product(self):
+        self.save()
+        try:
+            OrderedProduct.full_clean(self)
+        except ValidationError as e:
+            self.delete()
+            raise e
+        return self
+
+    def delete_ordered_product(self):
+        OrderedProduct.objects.get(id=self.id).delete()
+
+    def decrease_quantity(self, qn):
+        self.quantity -= qn
+        self.save()
+        if (self.quantity <= 0):
+            self.delete_ordered_product()
+
+    def increase_quantity(self, qn):
+        self.quantity += qn
+
+    @staticmethod
+    def order(delivery, client, supplier_product, qn):
```

Comment:
maybe check whether qn is 0, since we agreed product with 0 quantity is not valid.
would suggest raising an error if qn is 0 or create other indication the instance hasnt been saved.

Reply:
if qn <= 0 and not only 0

Reply:
Okay